modules/spring/src/main/java/org/apache/ignite/IgniteClientSpringBean.java

```diff
@@ -0,0 +1,208 @@
+        throws ClientException
+    {
+        return cli.getOrCreateCacheAsync(cfg);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <K, V> ClientCache<K, V> cache(String name) {
+        return cli.cache(name);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Collection<String> cacheNames() throws ClientException {
+        return cli.cacheNames();
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteClientFuture<Collection<String>> cacheNamesAsync() throws ClientException {
+        return cli.cacheNamesAsync();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void destroyCache(String name) throws ClientException {
+        cli.destroyCache(name);
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteClientFuture<Void> destroyCacheAsync(String name) throws ClientException {
+        return cli.destroyCacheAsync(name);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <K, V> ClientCache<K, V> createCache(String name) throws ClientException {
+        return cli.createCache(name);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <K, V> IgniteClientFuture<ClientCache<K, V>> createCacheAsync(String name) throws ClientException {
+        return cli.createCacheAsync(name);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <K, V> ClientCache<K, V> createCache(ClientCacheConfiguration cfg) throws ClientException {
+        return cli.createCache(cfg);
+    }
+
+    /** {@inheritDoc} */
+    @Override public <K, V> IgniteClientFuture<ClientCache<K, V>> createCacheAsync(ClientCacheConfiguration cfg)
+        throws ClientException
+    {
+        return cli.createCacheAsync(cfg);
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteBinary binary() {
+        return cli.binary();
+    }
+
+    /** {@inheritDoc} */
+    @Override public FieldsQueryCursor<List<?>> query(SqlFieldsQuery qry) {
+        return cli.query(qry);
+    }
+
+    /** {@inheritDoc} */
+    @Override public ClientTransactions transactions() {
+        return cli.transactions();
+    }
+
+    /** {@inheritDoc} */
+    @Override public ClientCompute compute() {
+        return cli.compute();
+    }
+
+    /** {@inheritDoc} */
+    @Override public ClientCompute compute(ClientClusterGroup grp) {
+        return cli.compute(grp);
+    }
+
+    /** {@inheritDoc} */
+    @Override public ClientCluster cluster() {
+        return cli.cluster();
+    }
+
+    /** {@inheritDoc} */
+    @Override public ClientServices services() {
+        return cli.services();
+    }
+
+    /** {@inheritDoc} */
+    @Override public ClientServices services(ClientClusterGroup grp) {
+        return cli.services(grp);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() throws Exception {
+        cli.close();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void afterSingletonsInstantiated() {
```

Comment:
Please add a test where both IgniteSpringBean and IgniteClientSpringBean defined in the XML. We should ensure that IgniteSpringBean initialized before IgniteClientSpringBean and client can connect to Ignite node without error. Perhaps we should move from SmartInitializingSingleton to SmartLifecycle and define different phases for IgniteSpringBean and IgniteClientSpringBean.

Reply:
I updated the PR and declared different phases for thin client and node initialization. It seems that using of SmartLifecycle is the only way to ensure that the Ignite node and Ignite thin client beans are initialized in the correct order regardless of the declaration order.